rockchip: add support for NanoPi R4S (WIP)

mj22226 opened the pull request.

1715173329 commented:
comes with error
```
cp -fpR /home/ctcgfw/op-dev/build_dir/target-aarch64_generic_musl/linux-rockchip_armv8/friendlyarm_nanopi-r4s-kernel.bin /home/ctcgfw/op-dev/build_dir/target-aarch64_generic_musl/linux-rockchip_armv8/tmp/openwrt-rockchip-armv8-friendlyarm_nanopi-r4s-ext4-sysupgrade.img.gz.boot/kernel.img
# Make an U-boot image and copy it to the boot partition
mkimage -A arm -O linux -T script -C none -a 0 -e 0 -d bootscript/.bootscript /home/ctcgfw/op-dev/build_dir/target-aarch64_generic_musl/linux-rockchip_armv8/tmp/openwrt-rockchip-armv8-friendlyarm_nanopi-r4s-ext4-sysupgrade.img.gz.boot/boot.scr
mkimage: Can't stat bootscript/.bootscript: No such file or directory
Makefile:67: recipe for target '/home/ctcgfw/op-dev/build_dir/target-aarch64_generic_musl/linux-rockchip_armv8/tmp/openwrt-rockchip-armv8-friendlyarm_nanopi-r4s-ext4-sysupgrade.img.gz' failed
make[5]: *** [/home/ctcgfw/op-dev/build_dir/target-aarch64_generic_musl/linux-rockchip_armv8/tmp/openwrt-rockchip-armv8-friendlyarm_nanopi-r4s-ext4-sysupgrade.img.gz] Error 1
```

mj22226 commented:
> comes with error
> 
> ```
> cp -fpR /home/ctcgfw/op-dev/build_dir/target-aarch64_generic_musl/linux-rockchip_armv8/friendlyarm_nanopi-r4s-kernel.bin /home/ctcgfw/op-dev/build_dir/target-aarch64_generic_musl/linux-rockchip_armv8/tmp/openwrt-rockchip-armv8-friendlyarm_nanopi-r4s-ext4-sysupgrade.img.gz.boot/kernel.img
> # Make an U-boot image and copy it to the boot partition
> mkimage -A arm -O linux -T script -C none -a 0 -e 0 -d bootscript/.bootscript /home/ctcgfw/op-dev/build_dir/target-aarch64_generic_musl/linux-rockchip_armv8/tmp/openwrt-rockchip-armv8-friendlyarm_nanopi-r4s-ext4-sysupgrade.img.gz.boot/boot.scr
> mkimage: Can't stat bootscript/.bootscript: No such file or directory
> Makefile:67: recipe for target '/home/ctcgfw/op-dev/build_dir/target-aarch64_generic_musl/linux-rockchip_armv8/tmp/openwrt-rockchip-armv8-friendlyarm_nanopi-r4s-ext4-sysupgrade.img.gz' failed
> make[5]: *** [/home/ctcgfw/op-dev/build_dir/target-aarch64_generic_musl/linux-rockchip_armv8/tmp/openwrt-rockchip-armv8-friendlyarm_nanopi-r4s-ext4-sysupgrade.img.gz] Error 1
> ```

I look like u boot-rockchip didn't build, are you building with in a clean source?, if not you may try that, it build fine here.

1715173329 commented:
> I look like u boot-rockchip didn't build, are you building with in a clean source?, if not you may try that, it build fine here.

However, it works fine for me after commit 5ad0e8e.

mj22226 commented:
> Maybe we can drop this line as it can be defined automatically via `Makefile`.

Since I don't have the device yet, can't test that right now.

1715173329 commented:
> Since I don't have the device yet, can't test that right now.

I've tested and it works fine.

mj22226 commented:
There are 3 DDR3 options for rk3399 in u boot:
```
rk3399-sdram-ddr3-1333.dtsi
rk3399-sdram-ddr3-1600.dtsi
rk3399-sdram-ddr3-1866.dtsi
```
Try  with ddr3-1600

1715173329 commented:
> There are 3 DDR3 options for rk3399 in u boot:
> 
> ```
> rk3399-sdram-ddr3-1333.dtsi
> rk3399-sdram-ddr3-1600.dtsi
> rk3399-sdram-ddr3-1866.dtsi
> ```
> 
> Try with ddr3-1600 and drop `-CONFIG_RAM_RK3399_LPDDR4=y` in
> 
> https://github.com/mj22226/openwrt/blob/1d33de9993c7710a8bac2111d55e0c2821fbbd24/package/boot/uboot-rockchip/patches/200-uboot-rockchip-add-support-for-NanoPi-R4S-v11.patch#L187

All right, I will try later

umbrellary commented:
I tried it, and after deleting CONFIG_RAM_RK3399_LPDDR4=y & rk3399-sdram-lpddr4-100.dtsi, it can be booted.
But I also found two other problems:
1. The order of LAN and WAN is reversed
2. reboot will fail and can only be recovered by hardware reset

1715173329 commented:
> 1. The order of LAN and WAN is reversed

Try to apply this patch:
```patch
diff --git a/target/linux/rockchip/armv8/base-files/etc/board.d/02_network b/target/linux/rockchip/armv8/base-files/etc/board.d/02_network
index 6db0680db2..75f43f2dc8 100755
--- a/target/linux/rockchip/armv8/base-files/etc/board.d/02_network
+++ b/target/linux/rockchip/armv8/base-files/etc/board.d/02_network
@@ -8,9 +8,7 @@ rockchip_setup_interfaces()
 	local board="$1"
 
 	case "$board" in
-	friendlyarm,nanopi-r2s)
-		ucidef_set_interfaces_lan_wan 'eth1' 'eth0'
-		;;
+	friendlyarm,nanopi-r2s|\
 	friendlyarm,nanopi-r4s)
 		ucidef_set_interfaces_lan_wan 'eth1' 'eth0'
 		;;
@@ -28,7 +26,8 @@ rockchip_setup_macs()
 	local label_mac=""
 
 	case "$board" in
-	friendlyarm,nanopi-r2s)
+	friendlyarm,nanopi-r2s|\
+	friendlyarm,nanopi-r4s)
 		wan_mac=$(macaddr_random)
 		lan_mac=$(macaddr_add "$wan_mac" +1)
 		;;
```

1715173329 commented:
> 2. reboot will fail and can only be recovered by hardware reset

howerver, `reboot` works fine for me.

mj22226 commented:
> After `CONFIG_RAM_RK3399_LPDDR4=y` & `rk3399-sdram-lpddr4-100.dtsi` being dropped, I can finally boot with `rk3399-sdram-ddr3-1600.dtsi`
> 
> So maybe we need separate them into two packages?

Until we can find a fix we can have the 2 versions separated. I added the 1gb ddr3 version , tested only for build, Can you confirm that boots?.

1715173329 commented:
> Until we can find a fix we can have the 2 versions separated. I added the 1gb ddr3 version , tested only for build, Can you confirm that boots?.

It certainly works, but there're so many contents we can reuse to make it clean.
I'll do some tests and open a PR here.

1715173329 commented:
btw, I've added LEDs support for it, and you may take a look
https://github.com/1715173329/openwrt/commit/1a573d4e8dbb41d3aa9747645c100f23c15f0129

mj22226 commented:
> btw, I've added LEDs support for it, and you may take a look
> [1715173329/openwrt@1a573d4](https://github.com/1715173329/openwrt/commit/1a573d4e8dbb41d3aa9747645c100f23c15f0129)

LGTM

> > Until we can find a fix we can have the 2 versions separated. I added the 1gb ddr3 version , tested only for build, Can you confirm that boots?.
> 
> It certainly works, but there're so many contents we can reuse to make it clean.
> I'll do some tests and open a PR here.

Yes, I know , a clean up is needed to make it formal.

umbrellary commented:
reboot faild, uart output:

root@OpenWrt:/# [   29.175102] br-lan: port 1(eth1) entered disabled state
[   29.181060] device eth1 left promiscuous mode
[   29.181552] br-lan: port 1(eth1) entered disabled state
[   29.232463] r8169 0000:01:00.0 eth1: Link is Down
[   29.340822] rk_gmac-dwmac fe300000.ethernet eth0: Link is Down
[   31.707771] vcc1v8_s3: supplied by vcc_1v8
[   31.708277] vbus_typec: disabling
[   33.455861] reboot: Restarting system

U-Boot TPL 2020.07 (Dec 11 2020 - 06:32:18)
Channel 0: DDR3, 800MHz
BW=32 Col=10 Bk=8 CS0 Row=15 CS=1 Die BW=16 Size=1024MB
no stride
no stride
Trying to boot from BOOTROM
Returning to boot ROM...

U-Boot SPL 2020.07 (Dec 11 2020 - 06:32:18 +0000)
Trying to boot from MMC1
mmc_load_image_raw_sector: mmc block read error
Trying to boot from MMC1
mmc_load_image_raw_sector: mmc block read error
SPL: failed to boot from all boot devices
### ERROR ### Please RESET the board ###

gitgayhub commented:
4GB version confirmed working, but can't reboot

mj22226 commented:
> 4GB version confirmed working, but can't reboot

Other user has confirmed reboot working properly, you can also try a build with the new kernel 5.10
https://github.com/mj22226/openwrt/commits/rockchip-5-10

gitgayhub commented:
> > 4GB version confirmed working, but can't reboot
> 
> Other user has confirmed reboot working properly, you can also try a build with the new kernel 5.10
> https://github.com/mj22226/openwrt/commits/rockchip-5-10

Thanks, will try it later this week. But I prefer the normal 5.8 kernel.

Have you got your device?

Also, Is it possible to add support for 19.07.5 stable release? Thanks

1715173329 commented:
https://github.com/1715173329/openwrt/commit/07593c5a9d362674f1c397ebeec303c210877c72

I've tried another way to make a workaround, maybe it's more tiny

greekstreet commented:
No architecture detected on LuCI overview.
![22EEEC90-10BC-4696-8302-8B3B73695984](https://user-images.githubusercontent.com/4527080/102680761-3f9dd280-41f6-11eb-8568-06329dbb71a3.jpeg)

greekstreet commented:
Another problem found. One of the two USB ports doesn't work. Tried official friendlywrt ROM. Both of them is working. Here's the dmesg. 
`[    0.000000] Booting Linux on physical CPU 0x0000000000 [0x410fd034]
[    0.000000] Linux version 5.4.82 (yding@zbox) (gcc version 8.4.0 (OpenWrt GCC 8.4.0 r15249-55ae604be1)) #0 SMP PREEMPT Tue Dec 15 09:24:25 2020
[    0.000000] Machine model: FriendlyElec NanoPi R4S
[    0.000000] earlycon: uart8250 at MMIO32 0x00000000ff1a0000 (options '')
[    0.000000] printk: bootconsole [uart8250] enabled
[    0.000000] cma: Reserved 8 MiB at 0x00000000f7800000
[    0.000000] On node 0 totalpages: 1015296
[    0.000000]   DMA32 zone: 15864 pages used for memmap
[    0.000000]   DMA32 zone: 0 pages reserved
[    0.000000]   DMA32 zone: 1015296 pages, LIFO batch:63
[    0.000000] psci: probing for conduit method from DT.
[    0.000000] psci: PSCIv1.1 detected in firmware.
[    0.000000] psci: Using standard PSCI v0.2 function IDs
[    0.000000] psci: MIGRATE_INFO_TYPE not supported.
[    0.000000] psci: SMC Calling Convention v1.0
[    0.000000] percpu: Embedded 21 pages/cpu s45848 r8192 d31976 u86016
[    0.000000] pcpu-alloc: s45848 r8192 d31976 u86016 alloc=21*4096
[    0.000000] pcpu-alloc: [0] 0 [0] 1 [0] 2 [0] 3 [0] 4 [0] 5 
[    0.000000] Detected VIPT I-cache on CPU0
[    0.000000] CPU features: detected: ARM erratum 845719
[    0.000000] CPU features: detected: GIC system register CPU interface
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 999432
[    0.000000] Kernel command line: console=ttyS2,1500000 earlycon=uart8250,mmio32,0xff1a0000 root=PARTUUID=5452574f-02 rw rootwait
[    0.000000] Dentry cache hash table entries: 524288 (order: 10, 4194304 bytes, linear)
[    0.000000] Inode-cache hash table entries: 262144 (order: 9, 2097152 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 3959892K/4061184K available (7358K kernel code, 532K rwdata, 2288K rodata, 1600K init, 703K bss, 93100K reserved, 8192K cma-reserved)
[    0.000000] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=6, Nodes=1
[    0.000000] rcu: Preemptible hierarchical RCU implementation.
[    0.000000] rcu: 	RCU event tracing is enabled.
[    0.000000] rcu: 	RCU restricting CPUs from NR_CPUS=256 to nr_cpu_ids=6.
[    0.000000] 	Tasks RCU enabled.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 25 jiffies.
[    0.000000] rcu: Adjusting geometry for rcu_fanout_leaf=16, nr_cpu_ids=6
[    0.000000] NR_IRQS: 64, nr_irqs: 64, preallocated irqs: 0
[    0.000000] GICv3: GIC: Using split EOI/Deactivate mode
[    0.000000] GICv3: 256 SPIs implemented
[    0.000000] GICv3: 0 Extended SPIs implemented
[    0.000000] GICv3: Distributor has no Range Selector support
[    0.000000] GICv3: 16 PPIs implemented
[    0.000000] GICv3: no VLPI support, no direct LPI support
[    0.000000] GICv3: CPU0: found redistributor 0 region 0:0x00000000fef00000
[    0.000000] ITS [mem 0xfee20000-0xfee3ffff]
[    0.000000] ITS@0x00000000fee20000: allocated 65536 Devices @f2080000 (flat, esz 8, psz 64K, shr 0)
[    0.000000] ITS: using cache flushing for cmd queue
[    0.000000] GICv3: using LPI property table @0x00000000f2040000
[    0.000000] GIC: using cache flushing for LPI property table
[    0.000000] GICv3: CPU0: using allocated LPI pending table @0x00000000f2050000
[    0.000000] GICv3: GIC: PPI partition interrupt-partition-0[0] { /cpus/cpu@0[0] /cpus/cpu@1[1] /cpus/cpu@2[2] /cpus/cpu@3[3] }
[    0.000000] GICv3: GIC: PPI partition interrupt-partition-1[1] { /cpus/cpu@100[4] /cpus/cpu@101[5] }
[    0.000000] random: get_random_bytes called from start_kernel+0x29c/0x39c with crng_init=0
[    0.000000] arch_timer: cp15 timer(s) running at 24.00MHz (phys).
[    0.000000] clocksource: arch_sys_counter: mask: 0xffffffffffffff max_cycles: 0x588fe9dc0, max_idle_ns: 440795202592 ns
[    0.000006] sched_clock: 56 bits at 24MHz, resolution 41ns, wraps every 4398046511097ns
[    0.001886] Console: colour dummy device 80x25
[    0.002351] Calibrating delay loop (skipped), value calculated using timer frequency.. 48.00 BogoMIPS (lpj=96000)
[    0.003333] pid_max: default: 32768 minimum: 301
[    0.003990] Mount-cache hash table entries: 8192 (order: 4, 65536 bytes, linear)
[    0.004713] Mountpoint-cache hash table entries: 8192 (order: 4, 65536 bytes, linear)
[    0.007440] ASID allocator initialised with 32768 entries
[    0.008073] rcu: Hierarchical SRCU implementation.
[    0.008818] Platform MSI: interrupt-controller@fee20000 domain created
[    0.009869] PCI/MSI: /interrupt-controller@fee00000/interrupt-controller@fee20000 domain created
[    0.011303] smp: Bringing up secondary CPUs ...
[    0.012334] Detected VIPT I-cache on CPU1
[    0.012371] GICv3: CPU1: found redistributor 1 region 0:0x00000000fef20000
[    0.012387] GICv3: CPU1: using allocated LPI pending table @0x00000000f2060000
[    0.012434] CPU1: Booted secondary processor 0x0000000001 [0x410fd034]
[    0.013088] Detected VIPT I-cache on CPU2
[    0.013116] GICv3: CPU2: found redistributor 2 region 0:0x00000000fef40000
[    0.013130] GICv3: CPU2: using allocated LPI pending table @0x00000000f2070000
[    0.013161] CPU2: Booted secondary processor 0x0000000002 [0x410fd034]
[    0.013802] Detected VIPT I-cache on CPU3
[    0.013828] GICv3: CPU3: found redistributor 3 region 0:0x00000000fef60000
[    0.013841] GICv3: CPU3: using allocated LPI pending table @0x00000000f2100000
[    0.013871] CPU3: Booted secondary processor 0x0000000003 [0x410fd034]
[    0.014509] CPU features: detected: EL2 vector hardening
[    0.014518] ARM_SMCCC_ARCH_WORKAROUND_1 missing from firmware
[    0.014526] Detected PIPT I-cache on CPU4
[    0.014556] GICv3: CPU4: found redistributor 100 region 0:0x00000000fef80000
[    0.014569] GICv3: CPU4: using allocated LPI pending table @0x00000000f2110000
[    0.014604] CPU4: Booted secondary processor 0x0000000100 [0x410fd082]
[    0.015250] Detected PIPT I-cache on CPU5
[    0.015276] GICv3: CPU5: found redistributor 101 region 0:0x00000000fefa0000
[    0.015288] GICv3: CPU5: using allocated LPI pending table @0x00000000f2120000
[    0.015315] CPU5: Booted secondary processor 0x0000000101 [0x410fd082]
[    0.015413] smp: Brought up 1 node, 6 CPUs
[    0.028594] SMP: Total of 6 processors activated.
[    0.029047] CPU features: detected: 32-bit EL0 Support
[    0.029537] CPU features: detected: CRC32 instructions
[    0.041596] CPU: All CPU(s) started at EL2
[    0.042029] alternatives: patching kernel code
[    0.058166] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
[    0.059145] futex hash table entries: 2048 (order: 5, 131072 bytes, linear)
[    0.060168] pinctrl core: initialized pinctrl subsystem
[    0.061259] NET: Registered protocol family 16
[    0.063958] DMA: preallocated 256 KiB pool for atomic allocations
[    0.065501] cpuidle: using governor menu
[    0.066085] Serial: AMBA PL011 UART driver
[    0.093216] HugeTLB registered 1.00 GiB page size, pre-allocated 0 pages
[    0.093878] HugeTLB registered 32.0 MiB page size, pre-allocated 0 pages
[    0.094535] HugeTLB registered 2.00 MiB page size, pre-allocated 0 pages
[    0.095207] HugeTLB registered 64.0 KiB page size, pre-allocated 0 pages
[    0.099197] vcc3v0_sd: supplied by vcc3v3_sys
[    0.100319] iommu: Default domain type: Translated 
[    0.102757] SCSI subsystem initialized
[    0.103312] usbcore: registered new interface driver usbfs
[    0.103884] usbcore: registered new interface driver hub
[    0.104480] usbcore: registered new device driver usb
[    0.105526] workqueue: max_active 576 requested for napi_workq is out of range, clamping between 1 and 512
[    0.107039] clocksource: Switched to clocksource arch_sys_counter
[    0.107816] VFS: Disk quotas dquot_6.6.0
[    0.108256] VFS: Dquot-cache hash table entries: 512 (order 0, 4096 bytes)
[    0.112109] thermal_sys: Registered thermal governor 'step_wise'
[    0.112113] thermal_sys: Registered thermal governor 'power_allocator'
[    0.113216] NET: Registered protocol family 2
[    0.114686] tcp_listen_portaddr_hash hash table entries: 2048 (order: 3, 32768 bytes, linear)
[    0.115611] TCP established hash table entries: 32768 (order: 6, 262144 bytes, linear)
[    0.116727] TCP bind hash table entries: 32768 (order: 7, 524288 bytes, linear)
[    0.118099] TCP: Hash tables configured (established 32768 bind 32768)
[    0.118854] UDP hash table entries: 2048 (order: 4, 65536 bytes, linear)
[    0.119654] UDP-Lite hash table entries: 2048 (order: 4, 65536 bytes, linear)
[    0.120624] NET: Registered protocol family 1
[    0.121073] PCI: CLS 0 bytes, default 64
[    0.122716] workingset: timestamp_bits=46 max_order=20 bucket_order=0
[    0.129217] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.129792] jffs2: version 2.2 (NAND) (SUMMARY) (ZLIB) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.132118] Block layer SCSI generic (bsg) driver version 0.4 loaded (major 252)
[    0.133010] io scheduler mq-deadline registered
[    0.140599] rockchip-pcie f8000000.pcie: no vpcie12v regulator found
[    0.141310] rockchip-pcie f8000000.pcie: no vpcie1v8 regulator found
[    0.141948] rockchip-pcie f8000000.pcie: no vpcie0v9 regulator found
[    0.179830] rockchip-pcie f8000000.pcie: host bridge /pcie@f8000000 ranges:
[    0.180536] rockchip-pcie f8000000.pcie:   MEM 0xfa000000..0xfbdfffff -> 0xfa000000
[    0.181295] rockchip-pcie f8000000.pcie:    IO 0xfbe00000..0xfbefffff -> 0xfbe00000
[    0.182307] rockchip-pcie f8000000.pcie: PCI host bridge to bus 0000:00
[    0.182961] pci_bus 0000:00: root bus resource [bus 00-1f]
[    0.183527] pci_bus 0000:00: root bus resource [mem 0xfa000000-0xfbdfffff]
[    0.184205] pci_bus 0000:00: root bus resource [io  0x0000-0xfffff] (bus address [0xfbe00000-0xfbefffff])
[    0.185170] pci 0000:00:00.0: [1d87:0100] type 01 class 0x060400
[    0.185880] pci 0000:00:00.0: supports D1
[    0.186277] pci 0000:00:00.0: PME# supported from D0 D1 D3hot
[    0.191349] pci 0000:00:00.0: bridge configuration invalid ([bus 00-00]), reconfiguring
[    0.192308] pci 0000:01:00.0: [10ec:8168] type 00 class 0x020000
[    0.193004] pci 0000:01:00.0: reg 0x10: initial BAR value 0x00000000 invalid
[    0.193696] pci 0000:01:00.0: reg 0x10: [io  size 0x0100]
[    0.194303] pci 0000:01:00.0: reg 0x18: [mem 0x00000000-0x00000fff 64bit]
[    0.195017] pci 0000:01:00.0: reg 0x20: [mem 0x00000000-0x00003fff 64bit]
[    0.195745] pci 0000:01:00.0: Upstream bridge's Max Payload Size set to 128 (was 256, max 256)
[    0.196600] pci 0000:01:00.0: Max Payload Size set to 128 (was 128, max 128)
[    0.197623] pci 0000:01:00.0: supports D1 D2
[    0.198045] pci 0000:01:00.0: PME# supported from D0 D1 D2 D3hot D3cold
[    0.215374] pci_bus 0000:01: busn_res: [bus 01-1f] end is updated to 01
[    0.216044] pci 0000:00:00.0: BAR 8: assigned [mem 0xfa000000-0xfa0fffff]
[    0.216717] pci 0000:01:00.0: BAR 4: assigned [mem 0xfa000000-0xfa003fff 64bit]
[    0.217474] pci 0000:01:00.0: BAR 2: assigned [mem 0xfa004000-0xfa004fff 64bit]
[    0.218229] pci 0000:01:00.0: BAR 0: no space for [io  size 0x0100]
[    0.218844] pci 0000:01:00.0: BAR 0: failed to assign [io  size 0x0100]
[    0.219509] pci 0000:00:00.0: PCI bridge to [bus 01]
[    0.220003] pci 0000:00:00.0:   bridge window [mem 0xfa000000-0xfa0fffff]
[    0.220803] pcieport 0000:00:00.0: enabling device (0000 -> 0002)
[    0.221629] pcieport 0000:00:00.0: PME: Signaling with IRQ 224
[    0.222425] pcieport 0000:00:00.0: AER: enabled with IRQ 224
[    0.224927] dma-pl330 ff6d0000.dma-controller: Loaded driver for PL330 DMAC-241330
[    0.225675] dma-pl330 ff6d0000.dma-controller: 	DBUFF-32x8bytes Num_Chans-6 Num_Peri-12 Num_Events-12
[    0.227649] dma-pl330 ff6e0000.dma-controller: Loaded driver for PL330 DMAC-241330
[    0.228396] dma-pl330 ff6e0000.dma-controller: 	DBUFF-128x8bytes Num_Chans-8 Num_Peri-20 Num_Events-16
[    0.230904] Serial: 8250/16550 driver, 4 ports, IRQ sharing enabled
[    0.232794] ff180000.serial: ttyS0 at MMIO 0xff180000 (irq = 37, base_baud = 1500000) is a 16550A
[    0.233779] serial serial0: tty port ttyS0 registered
[    0.234653] ff1a0000.serial: ttyS2 at MMIO 0xff1a0000 (irq = 38, base_baud = 1500000) is a 16550A
[    0.235695] printk: console [ttyS2] enabled
[    0.236472] printk: bootconsole [uart8250] disabled
[    0.238757] cacheinfo: Unable to detect cache hierarchy for CPU 0
[    0.243739] loop: module loaded
[    0.244026] mtip32xx Version 1.3.1
[    0.246209] libphy: Fixed MDIO Bus: probed
[    0.247558] rk_gmac-dwmac fe300000.ethernet: IRQ eth_wake_irq not found
[    0.248146] rk_gmac-dwmac fe300000.ethernet: IRQ eth_lpi not found
[    0.248792] rk_gmac-dwmac fe300000.ethernet: PTP uses main clock
[    0.249356] rk_gmac-dwmac fe300000.ethernet: phy regulator is not available yet, deferred probing
[    0.251855] dwc3 fe800000.usb: Configuration mismatch. dr_mode forced to host
[    0.262704] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    0.263298] ehci-pci: EHCI PCI platform driver
[    0.263734] ehci-platform: EHCI generic platform driver
[    0.266369] ehci-platform fe380000.usb: EHCI Host Controller
[    0.266887] ehci-platform fe380000.usb: new USB bus registered, assigned bus number 1
[    0.267690] ehci-platform fe380000.usb: irq 29, io mem 0xfe380000
[    0.283061] ehci-platform fe380000.usb: USB 2.0 started, EHCI 1.00
[    0.284037] hub 1-0:1.0: USB hub found
[    0.284396] hub 1-0:1.0: 1 port detected
[    0.287173] ehci-platform fe3c0000.usb: EHCI Host Controller
[    0.287687] ehci-platform fe3c0000.usb: new USB bus registered, assigned bus number 2
[    0.288457] ehci-platform fe3c0000.usb: irq 31, io mem 0xfe3c0000
[    0.303060] ehci-platform fe3c0000.usb: USB 2.0 started, EHCI 1.00
[    0.303992] hub 2-0:1.0: USB hub found
[    0.304354] hub 2-0:1.0: 1 port detected
[    0.305163] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    0.305722] ohci-platform: OHCI generic platform driver
[    0.306331] ohci-platform fe3a0000.usb: Generic Platform OHCI controller
[    0.306932] ohci-platform fe3a0000.usb: new USB bus registered, assigned bus number 3
[    0.307720] ohci-platform fe3a0000.usb: irq 30, io mem 0xfe3a0000
[    0.371594] hub 3-0:1.0: USB hub found
[    0.371956] hub 3-0:1.0: 1 port detected
[    0.372681] ohci-platform fe3e0000.usb: Generic Platform OHCI controller
[    0.373289] ohci-platform fe3e0000.usb: new USB bus registered, assigned bus number 4
[    0.374088] ohci-platform fe3e0000.usb: irq 32, io mem 0xfe3e0000
[    0.435588] hub 4-0:1.0: USB hub found
[    0.435947] hub 4-0:1.0: 1 port detected
[    0.437226] xhci-hcd xhci-hcd.0.auto: xHCI Host Controller
[    0.437726] xhci-hcd xhci-hcd.0.auto: new USB bus registered, assigned bus number 5
[    0.438527] xhci-hcd xhci-hcd.0.auto: hcc params 0x0220fe64 hci version 0x110 quirks 0x0000000002010010
[    0.439404] xhci-hcd xhci-hcd.0.auto: irq 225, io mem 0xfe800000
[    0.440467] hub 5-0:1.0: USB hub found
[    0.440830] hub 5-0:1.0: 1 port detected
[    0.441421] xhci-hcd xhci-hcd.0.auto: xHCI Host Controller
[    0.441914] xhci-hcd xhci-hcd.0.auto: new USB bus registered, assigned bus number 6
[    0.442593] xhci-hcd xhci-hcd.0.auto: Host supports USB 3.0 SuperSpeed
[    0.443236] usb usb6: We don't know the algorithms for LPM for this host, disabling LPM.
[    0.444321] hub 6-0:1.0: USB hub found
[    0.444680] hub 6-0:1.0: 1 port detected
[    0.445394] xhci-hcd xhci-hcd.1.auto: xHCI Host Controller
[    0.445892] xhci-hcd xhci-hcd.1.auto: new USB bus registered, assigned bus number 7
[    0.446689] xhci-hcd xhci-hcd.1.auto: hcc params 0x0220fe64 hci version 0x110 quirks 0x0000000002010010
[    0.447563] xhci-hcd xhci-hcd.1.auto: irq 226, io mem 0xfe900000
[    0.448641] hub 7-0:1.0: USB hub found
[    0.449001] hub 7-0:1.0: 1 port detected
[    0.449591] xhci-hcd xhci-hcd.1.auto: xHCI Host Controller
[    0.450084] xhci-hcd xhci-hcd.1.auto: new USB bus registered, assigned bus number 8
[    0.450764] xhci-hcd xhci-hcd.1.auto: Host supports USB 3.0 SuperSpeed
[    0.451411] usb usb8: We don't know the algorithms for LPM for this host, disabling LPM.
[    0.452484] hub 8-0:1.0: USB hub found
[    0.452842] hub 8-0:1.0: 1 port detected
[    0.453595] usbcore: registered new interface driver usb-storage
[    0.454285] i2c /dev entries driver
[    0.456834] fan53555-regulator 0-0040: FAN53555 Option[8] Rev[1] Detected!
[    0.458699] vcc5v0_sys: supplied by vdd_5v
[    0.459144] vdd_cpu_b: supplied by vcc3v3_sys
[    0.460113] fan53555-regulator 0-0041: FAN53555 Option[8] Rev[1] Detected!
[    0.461924] vdd_gpu: supplied by vcc3v3_sys
[    0.462817] rk808 0-001b: chip id: 0x0
[    0.467145] rk808-regulator rk808-regulator: there is no dvs0 gpio
[    0.467711] rk808-regulator rk808-regulator: there is no dvs1 gpio
[    0.468999] vdd_center: supplied by vcc3v3_sys
[    0.469894] vdd_cpu_l: supplied by vcc3v3_sys
[    0.470468] vcc_ddr: supplied by vcc3v3_sys
[    0.471342] vcc_1v8: supplied by vcc3v3_sys
[    0.472898] vcc1v8_cam: supplied by vcc3v3_sys
[    0.474167] vcc3v0_touch: supplied by vcc3v3_sys
[    0.475598] vcc1v8_pmupll: supplied by vcc3v3_sys
[    0.477068] vcc_sdio: supplied by vcc3v3_sys
[    0.478324] vcca3v0_codec: supplied by vcc3v3_sys
[    0.479770] vcc_1v5: supplied by vcc3v3_sys
[    0.481037] vcca1v8_codec: supplied by vcc3v3_sys
[    0.482483] vcc_3v0: supplied by vcc3v3_sys
[    0.483100] vcc3v3_s3: supplied by vcc3v3_sys
[    0.483714] vcc3v3_s0: supplied by vcc3v3_sys
[    0.487264] rk808-rtc rk808-rtc: registered as rtc0
[    0.491926] energy_model: pd0: hertz/watts ratio non-monotonically decreasing: em_cap_state 1 >= em_cap_state0
[    0.496653] sdhci: Secure Digital Host Controller Interface driver
[    0.497200] sdhci: Copyright(c) Pierre Ossman
[    0.497584] Synopsys Designware Multimedia Card Interface Driver
[    0.498843] dwmmc_rockchip fe320000.dwmmc: IDMAC supports 32-bit address mode.
[    0.499019] random: fast init done
[    0.499514] dwmmc_rockchip fe320000.dwmmc: Using internal DMA controller.
[    0.500386] dwmmc_rockchip fe320000.dwmmc: Version ID is 270a
[    0.500936] dwmmc_rockchip fe320000.dwmmc: DW MMC controller at irq 28,32 bit host data width,256 deep fifo
[    0.501889] dwmmc_rockchip fe320000.dwmmc: Got CD GPIO
[    0.515493] mmc_host mmc0: Bus speed (slot 0) = 400000Hz (slot req 400000Hz, actual 400000HZ div = 0)
[    0.529635] sdhci-pltfm: SDHCI platform and OF driver helper
[    0.531179] ledtrig-cpu: registered to indicate activity on CPUs
[    0.531971] usbcore: registered new interface driver usbhid
[    0.532459] usbhid: USB HID core driver
[    0.533613] NET: Registered protocol family 10
[    0.534553] Segment Routing with IPv6
[    0.534908] NET: Registered protocol family 17
[    0.535332] 8021q: 802.1Q VLAN Support v1.8
[    0.553348] rk_gmac-dwmac fe300000.ethernet: IRQ eth_wake_irq not found
[    0.553929] rk_gmac-dwmac fe300000.ethernet: IRQ eth_lpi not found
[    0.554528] rk_gmac-dwmac fe300000.ethernet: PTP uses main clock
[    0.555175] rk_gmac-dwmac fe300000.ethernet: clock input or output? (input).
[    0.555792] rk_gmac-dwmac fe300000.ethernet: TX delay(0x28).
[    0.556287] rk_gmac-dwmac fe300000.ethernet: RX delay(0x11).
[    0.556787] rk_gmac-dwmac fe300000.ethernet: integrated PHY? (no).
[    0.557346] rk_gmac-dwmac fe300000.ethernet: cannot get clock clk_mac_speed
[    0.557981] rk_gmac-dwmac fe300000.ethernet: clock input from PHY
[    0.563537] rk_gmac-dwmac fe300000.ethernet: init for RGMII
[    0.564626] rk_gmac-dwmac fe300000.ethernet: User ID: 0x10, Synopsys ID: 0x35
[    0.565256] rk_gmac-dwmac fe300000.ethernet: 	DWMAC1000
[    0.565713] rk_gmac-dwmac fe300000.ethernet: DMA HW capability register supported
[    0.566365] rk_gmac-dwmac fe300000.ethernet: RX Checksum Offload Engine supported
[    0.567015] rk_gmac-dwmac fe300000.ethernet: COE Type 2
[    0.567486] rk_gmac-dwmac fe300000.ethernet: TX Checksum insertion supported
[    0.568100] rk_gmac-dwmac fe300000.ethernet: Wake-Up On Lan supported
[    0.568661] rk_gmac-dwmac fe300000.ethernet: Normal descriptors
[    0.569177] rk_gmac-dwmac fe300000.ethernet: Ring mode enabled
[    0.569686] rk_gmac-dwmac fe300000.ethernet: Enable RX Mitigation via HW Watchdog Timer
[    0.570384] rk_gmac-dwmac fe300000.ethernet: device MAC address 86:79:21:30:a1:7a
[    0.615160] mmc_host mmc0: Bus speed (slot 0) = 148500000Hz (slot req 150000000Hz, actual 148500000HZ div = 0)
[    0.631047] libphy: stmmac: probed
[    0.634720] rk808-rtc rk808-rtc: setting system clock to 2013-01-18T08:50:09 UTC (1358499009)
[    0.636749] ttyS2 - failed to request DMA
[    0.637179] Waiting for root device PARTUUID=5452574f-02...
[    0.790774] dwmmc_rockchip fe320000.dwmmc: Successfully tuned phase to 213
[    0.791461] mmc0: new ultra high speed SDR104 SDHC card at address aaaa
[    0.793607] mmcblk0: mmc0:aaaa SC32G 29.7 GiB 
[    0.798734]  mmcblk0: p1 p2
[    0.835744] EXT4-fs (mmcblk0p2): mounted filesystem without journal. Opts: (null)
[    0.836439] VFS: Mounted root (ext4 filesystem) on device 179:2.
[    0.837754] Freeing unused kernel memory: 1600K
[    0.843298] Run /sbin/init as init process
[    0.886193] init: Console is alive
[    1.020497] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    1.030245] ehci-fsl: Freescale EHCI Host controller driver
[    1.034524] usbcore: registered new interface driver uas
[    1.035151] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    1.039544] init: - preinit -
[    1.126205] random: jshn: uninitialized urandom read (4 bytes read)
[    1.147659] random: jshn: uninitialized urandom read (4 bytes read)
[    1.164346] random: jshn: uninitialized urandom read (4 bytes read)
[    5.284272] mount_root: mounting /dev/root
[    5.305745] EXT4-fs (mmcblk0p2): re-mounted. Opts: (null)
[    5.306327] mount_root: loading kmods from internal overlay
[    5.319170] kmodloader: loading kernel modules from //etc/modules-boot.d/*
[    5.320828] kmodloader: done loading kernel modules from //etc/modules-boot.d/*
[    5.486359] block: attempting to load /etc/config/fstab
[    5.487301] block: unable to load configuration (fstab: Entry not found)
[    5.487906] block: no usable configuration
[    5.566684] EXT4-fs (mmcblk0p1): mounted filesystem without journal. Opts: (null)
[    5.578172] urandom-seed: Seed file not found (/etc/urandom.seed)
[    5.591566] procd: - early -
[    6.153480] procd: - ubus -
[    6.159856] urandom_read: 2 callbacks suppressed
[    6.159859] random: ubusd: uninitialized urandom read (4 bytes read)
[    6.204338] random: ubusd: uninitialized urandom read (4 bytes read)
[    6.205400] random: ubusd: uninitialized urandom read (4 bytes read)
[    6.207706] procd: - init -
[    6.293573] urngd: v1.0.2 started.
[    6.308630] random: crng init done
[    6.308934] random: 2 urandom warning(s) missed due to ratelimiting
[    6.311554] kmodloader: loading kernel modules from /etc/modules.d/*
[    6.338649] NET: Registered protocol family 38
[    6.345344] device-mapper: ioctl: 4.41.0-ioctl (2019-09-16) initialised: dm-devel@redhat.com
[    6.351184] Key type dns_resolver registered
[    6.360035] RPC: Registered named UNIX socket transport module.
[    6.360554] RPC: Registered udp transport module.
[    6.360965] RPC: Registered tcp transport module.
[    6.361375] RPC: Registered tcp NFSv4.1 backchannel transport module.
[    6.387571] Installing knfsd (copyright (C) 1996 okir@monad.swb.de).
[    6.396872] NFS: Registering the id_resolver key type
[    6.397328] Key type id_resolver registered
[    6.397693] Key type id_legacy registered
[    6.421314] cryptodev: driver 1.10 loaded.
[    6.442852] r8169 0000:01:00.0: enabling device (0000 -> 0002)
[    6.455486] r8169 0000:01:00.0: can't read MAC address, setting random one
[    6.461775] libphy: r8169: probed
[    6.463001] r8169 0000:01:00.0 eth1: RTL8168h/8111h, aa:74:da:25:6e:c4, XID 541, IRQ 236
[    6.463743] r8169 0000:01:00.0 eth1: jumbo features [frames: 9200 bytes, tx checksumming: ko]
[    6.466152] usbcore: registered new interface driver ums-alauda
[    6.467155] usbcore: registered new interface driver ums-cypress
[    6.468220] usbcore: registered new interface driver ums-datafab
[    6.469444] usbcore: registered new interface driver ums-freecom
[    6.470783] usbcore: registered new interface driver ums-isd200
[    6.472267] usbcore: registered new interface driver ums-jumpshot
[    6.473379] usbcore: registered new interface driver ums-karma
[    6.474438] usbcore: registered new interface driver ums-sddr09
[    6.475498] usbcore: registered new interface driver ums-sddr55
[    6.476527] usbcore: registered new interface driver ums-usbat
[    6.479187] wireguard: WireGuard 1.0.20201112 loaded. See www.wireguard.com for information.
[    6.480643] wireguard: Copyright (C) 2015-2019 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
[    6.485337] xt_time: kernel timezone is -0000
[    6.493996] PPP generic driver version 2.4.2
[    6.494840] NET: Registered protocol family 24
[    6.498948] kmodloader: done loading kernel modules from /etc/modules.d/*
[    7.043394] phy phy-ff770000.syscon:usb2-phy@e450.1: charger = USB_DCP_CHARGER
[   11.550104] Generic FE-GE Realtek PHY r8169-100:00: attached PHY driver [Generic FE-GE Realtek PHY] (mii_bus:phy_addr=r8169-100:00, irq=IGNORE)
[   11.660216] r8169 0000:01:00.0 eth1: Link is Down
[   11.660449] br-lan: port 1(eth1) entered blocking state
[   11.661096] br-lan: port 1(eth1) entered disabled state
[   11.661699] device eth1 entered promiscuous mode
[   11.666544] rk_gmac-dwmac fe300000.ethernet eth0: PHY [stmmac-0:01] driver [RTL8211E Gigabit Ethernet]
[   11.669886] rk_gmac-dwmac fe300000.ethernet eth0: No Safety Features support found
[   11.670573] rk_gmac-dwmac fe300000.ethernet eth0: PTP not supported by HW
[   11.671248] rk_gmac-dwmac fe300000.ethernet eth0: configuring for phy/rgmii link mode
[   14.681170] r8169 0000:01:00.0 eth1: Link is Up - 1Gbps/Full - flow control rx/tx
[   14.681906] br-lan: port 1(eth1) entered blocking state
[   14.682376] br-lan: port 1(eth1) entered forwarding state
[   14.683757] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   14.699968] rk_gmac-dwmac fe300000.ethernet eth0: Link is Up - 1Gbps/Full - flow control off
[   14.700773] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   19.956267] pppoe-wan: renamed from ppp0
[   19.975599] NFSD: Using /var/lib/nfs/v4recovery as the NFSv4 state recovery directory
[   19.976319] NFSD: Using legacy client tracking operations.
[   19.976805] NFSD: starting 10-second grace period (net f00000a0)
[   31.715414] vcc1v8_s3: supplied by vcc_1v8
[   31.715921] vbus_typec: disabling`

mj22226 commented:
> Another problem found. One of the two USB ports doesn't work. Tried official friendlywrt ROM. Both of them is working. Here's the dmesg.

Can't help much right now, because I don't have the r4s in hands and that it won't be until next year if you have any ideas or find  a fix please report it.

greekstreet commented:
> > Another problem found. One of the two USB ports doesn't work. Tried official friendlywrt ROM. Both of them is working. Here's the dmesg.
> 
> Can't help much right now, because I don't have the r4s in hands and that it won't be until next year if you have any ideas or find a fix please report it.

Compiled rockchip-5-10 branch, both USB port works as expected on my R4S 4GB version. The build seems stable. Running for a day without issues.

Some packages failed though. 
GNUTLS failed but can be fixed by upgrading nettle to 3.6.
bash and nfs-kernel-server failed and I just unchecked them. 

Thank you for your code. Merry Christmas!
